pkg/agent/apiserver/handlers/ovsflows/handler_test.go

```diff
@@ -169,48 +182,140 @@ func TestServiceFlows(t *testing.T) {
+			q := aqtest.NewMockAgentQuerier(ctrl)
 			ofc := oftest.NewMockClient(ctrl)
 			ovsctl := ovsctltest.NewMockOVSCtlClient(ctrl)
-			npq.EXPECT().GetNetworkPolicies(&querier.NetworkPolicyQueryFilter{SourceName: tc.name, Namespace: tc.namespace}).Return([]cpv1beta.NetworkPolicy{*testNetworkPolicy}).Times(1)
-			ofc.EXPECT().GetNetworkPolicyFlowKeys(tc.name, tc.namespace).Return(testFlowKeys).Times(1)
+			npFilter := &querier.NetworkPolicyQueryFilter{
+				SourceName: tc.name,
+				Namespace:  tc.namespace,
+				SourceType: tc.policyType,
+			}
+			q.EXPECT().GetNetworkPolicyInfoQuerier().Return(npq).Times(1)
+			npq.EXPECT().GetNetworkPolicies(npFilter).Return([]cpv1beta.NetworkPolicy{
+				{
+					SourceRef: &cpv1beta.NetworkPolicyReference{
+						Type:      tc.policyType,
+						Namespace: tc.namespace,
+						Name:      tc.name,
+					},
+				},
+			}).Times(1)
+			ofc.EXPECT().GetNetworkPolicyFlowKeys(tc.name, tc.namespace, tc.policyType).Return(testFlowKeys).Times(1)
 			q.EXPECT().GetOpenflowClient().Return(ofc).Times(1)
 			q.EXPECT().GetOVSCtlClient().Return(ovsctl).Times(len(testFlowKeys))
 			for i := range testFlowKeys {
 				ovsctl.EXPECT().DumpMatchedFlow(testFlowKeys[i]).Return(testDumpFlows[i], nil).Times(1)
 			}
-		} else {
-			npq.EXPECT().GetNetworkPolicies(&querier.NetworkPolicyQueryFilter{SourceName: tc.name, Namespace: tc.namespace}).Return(nil).Times(1)
-		}
+			runHTTPTest(t, &tc, q)
+		})
+	}
+}
 
-		runHTTPTest(t, &tc, q)
+func TestNetworkPolicyFlowsBadRequest(t *testing.T) {
+	testcases := []testCase{
+		{
+			testName:       "ACNP bad request",
+			name:           "acnp2",
+			policyType:     cpv1beta.AntreaClusterNetworkPolicy,
+			query:          "?networkpolicy=acnp2&type=ACNP&namespace=default",
+			expectedStatus: http.StatusBadRequest,
+		},
+		{
+			testName:       "ANNP bad request",
+			name:           "annp2",
+			policyType:     cpv1beta.AntreaNetworkPolicy,
+			query:          "?networkpolicy=annp2&type=ANNP",
+			expectedStatus: http.StatusBadRequest,
+		},
+		{
+			testName:       "Ambiguous query",
+			name:           "np-annp-same-name",
+			namespace:      "ns1",
+			query:          "?networkpolicy=np-annp-same-name&namespace=ns1",
+			expectedStatus: http.StatusBadRequest,
+		},
 	}
+	for i := range testcases {
+		tc := testcases[i]
+		t.Run(tc.testName, func(t *testing.T) {
+			ctrl := gomock.NewController(t)
+			q := aqtest.NewMockAgentQuerier(ctrl)
+			if tc.name == "np-annp-same-name" {
+				// Simulates an ambiguous query where more than one matching NP is returned
+				npq := queriertest.NewMockAgentNetworkPolicyInfoQuerier(ctrl)
+				q.EXPECT().GetNetworkPolicyInfoQuerier().Return(npq).Times(1)
+				npFilter := &querier.NetworkPolicyQueryFilter{
+					SourceName: tc.name,
+					Namespace:  tc.namespace,
+					SourceType: tc.policyType,
+				}
+				npq.EXPECT().GetNetworkPolicies(npFilter).Return([]cpv1beta.NetworkPolicy{*testNetworkPolicy, *testANNP}).Times(1)
+			}
```

Comment:
nit: `t.Run` should be generic and this test specific expectation should be moved to the test case itself by having a parameter like `expectedCalls func(recorder *queriertest.MockAgentQuerierMockRecorder)`

Reply:
Moved to its own test func